#Customer Review Sentiment Analysis using Machine Learning

This project analyzes customer reviews using machine learning techniques to extract sentiment and identify patterns that can provide business insights.

## Research Question

How can AI be used to analyze customer reviews and extract useful insights for decision-making?

In [25]:
!pip install datasets scikit-learn

import pandas as pd
from datasets import load_dataset
from sklearn.feature_extraction.text import TfidfVectorizer
from sklearn.linear_model import LogisticRegression

In [26]:
#Load Dataset
dataset = load_dataset("amazon_polarity")

data = dataset["train"].shuffle(seed=42).select(range(2000))

texts = data["content"]
labels = data["label"]

df = pd.DataFrame({"review": texts, "sentiment": labels})
df.head()

,review,sentiment
0,All the pretty people in this film. Even the R...,0
1,I know that Tom Robbins has a loyal following ...,0
2,This type of spaetzle maker is easier to manuv...,1
3,My cats seemed to be amused to work for their ...,1
4,I've had no problems running this with my Dell...,1


## Dataset

We use a subset of the Amazon Reviews dataset. Each review is labeled as positive or negative, allowing us to train a sentiment analysis model.

In [27]:
#Model Evaluation (Train-Test Split)

from sklearn.model_selection import train_test_split

X_train, X_test, y_train, y_test = train_test_split(
    df["review"], df["sentiment"], test_size=0.2, random_state=42
)

vectorizer = TfidfVectorizer(max_features=5000)
X_train_vec = vectorizer.fit_transform(X_train)
X_test_vec = vectorizer.transform(X_test)

model = LogisticRegression()
model.fit(X_train_vec, y_train)

preds = model.predict(X_test_vec)

from sklearn.metrics import accuracy_score
print("Accuracy:", accuracy_score(y_test, preds))

Accuracy: 0.775


To properly evaluate the model, we split the dataset into training and test sets.
This prevents the model from simply memorizing the data and provides a more realistic performance estimate.

In [28]:
# Show sample predictions
sample_df = pd.DataFrame({
    "review": X_test[:5],
    "true_label": y_test[:5],
    "predicted": preds[:5]
})

sample_df

,review,true_label,predicted
1860,"Again, a wonderful work of James Martin, SJ. T...",1,1
353,I was so very excited about the A Painted Hous...,1,0
1333,I upgraded to this from a cheap metal 1940's-s...,1,1
905,This book doesn't really have much of interest...,0,0
1289,"This story has everything! Comedy, compromise,...",1,1


## Results

The Logistic Regression model achieved an accuracy of approximately 0.77 on the test set.

Sample predictions show that the model can correctly classify both positive and negative reviews, although some errors remain.


## Interpretation

The model demonstrates that even simple machine learning techniques can effectively capture sentiment patterns in text data.

However, performance is limited compared to more advanced deep learning approaches, which could better capture context and nuance.

## Limitations

- Limited dataset size
- No deep learning models used
- No topic extraction beyond sentiment

## Future Work

- Use transformer models for better accuracy
- Perform topic modeling to identify key issues
- Apply to real business datasets